src/main/scala/rise/core/types/Constraints.scala

```diff
@@ -132,19 +131,18 @@ object Constraint {
                   substitute.natInType(n, `for`= nb, tb), n, preserve)
                 nTaSub ++ nTbSub ++ decomposed(
                     Seq(
-                      NatConstraint(n, na.asImplicit),
-                      NatConstraint(n, nb.asImplicit),
+                      NatConstraint(n, na),
+                      NatConstraint(n, nb),
                       TypeConstraint(nTa, nTb)
-                    ))
+                    ), preserve - na - nb)
               case ExplicitDependence.Off =>
                 val n = NatIdentifier(freshName("n"), isExplicit = true)
                 decomposed(
                   Seq(
-                    NatConstraint(n, na.asImplicit),
-                    NatConstraint(n, nb.asImplicit),
+                    NatConstraint(n, na),
+                    NatConstraint(n, nb),
                     TypeConstraint(ta, tb)
-                  )
-                )
+                  ), preserve - na - nb)
```

Comment:
`+ n`? same in other spots. I wonder if we could also alternatively directly apply the required substitutions (`na -> n` and `nb -> n`) where necessary and only generate 1 constraint instead of 3 here. Something like:
```scala
val n = NatIdentifier(freshName("n"))
decomposed(Seq(TypeConstraint(
  substitute.natsInType(ta, Map(na -> n)),
  substitute.natsInType(tb, Map(nb -> n))
))), preserve + n)
```

Reply:
Yup, I was just picking up on that, I will get back to you!